### PR TITLE
PAT-1207: add disruptions pages

### PR DESCRIPTION
**Pull request by connormaglynn**
- Add disruptions link to nominal home
- Add View Disruptions page
- Add Add Disruptions page
*overridden controls for ease of merges from Pathfinder

### CODE REVIEW

File: server/config/controlsAndInterventions/controls/getControls.js [hunk 1]
```diff
@@ -23,16 +24,17 @@ const {
...
   // Get a list of the unique needs attributes from the needs priority form data
-  const needsAttributes = R.flatten(getNeedsAttributesFromForm(needsForm))
+  // const needsAttributes = R.flatten(getNeedsAttributesFromForm(needsForm))
 
   // Get a list of controls for this offender
-  return allControls
-    .map(c => controlMatcher(needsAttributes, c, gender, riskBand, tactOffender, custodyOrCommunity))
-    .filter(c => !isNilOrEmpty(c))
+  return allControls.filter(c => !isNilOrEmpty(c))
+  // .map(c => controlMatcher(needsAttributes, c, gender, riskBand, tactOffender, custodyOrCommunity))
```
Comment (timharrison-moj) [inline server/config/controlsAndInterventions/controls/getControls.js:34]: Best to remove the commented lines that are not useful?

File: server/config/controlsAndInterventions/controls/getControls.js [hunk 2]
```diff
@@ -7,7 +8,7 @@ const {
   locationMatch,
   mandatoryTactMatch,
   mandatoryBandMatch,
-  getNeedsAttributesFromForm,
+  // getNeedsAttributesFromForm,
```
Comment (timharrison-moj) [inline server/config/controlsAndInterventions/controls/getControls.js:11]: Commented line can probably go?

**Review by timharrison-moj — COMMENTED** [review]
Generally looks good, though it will be a nightmare to merge PF changes into this. Bit of a rethink which might work out better for us - how about copying the controls routes, tests, form and config into other files called disruptions*.js? This would leave the original controls*.js files the same - and merging would not be affected. It might be an awful lot clearer too - at present we have files & functions relating to controls which are relly disruptions. A bit of rework, but I think its probably worth it for the separation and clarity it will bring for this PR and for any attempt at merging PF changes in over the top of them?

File: integration-tests/integration/disruptions/disruptions.spec.js [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+    cy.get('[id=ci-current-table]').should('contain.text', 'Analyse financial transactions')
+    cy.get('[id=ci-current-table]').should('contain.text', 'Area searching')
+    cy.get('[id=ci-current-table]').should('contain.text', 'SOC tasking meeting')
+    cy.get('[id=ci-current-table]').should('contain.text', 'James Stuart')
+    cy.get('[id=ci-current-table]').should('contain.text', 'Not started')
+
+    cy.contains('Sign out').click()
+  })
+})
```
Comment (timharrison-moj) [inline integration-tests/integration/disruptions/disruptions.spec.js:64]: This is fine as is.

File: migrations/20200824113100_cim_controls_constraints.js [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+      table.integer('needs_id').nullable().alter()
+    }),
+  ])
+
+exports.down = knex =>
+  Promise.all([
+    knex.schema.alterTable('cim_controls', table => {
+      table.integer('risk_band_id').notNullable().alter()
+      table.integer('needs_id').notNullable().alter()
+    }),
```
Comment (timharrison-moj) [inline migrations/20200824113100_cim_controls_constraints.js:14]: fine as is

File: migrations/20200824170800_active_controls_constraints.js [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+    }),
+  ])
+
+exports.down = knex =>
+  Promise.all([
+    knex.schema.alterTable('active_controls', table => {
+      table.integer('risk_band_id').notNullable().alter()
+      table.integer('needs_id').notNullable().alter()
+    }),
+  ])
```
Comment (timharrison-moj) [inline migrations/20200824170800_active_controls_constraints.js:15]: fine as is

File: server/config/controlsAndInterventions/controls/controls.js [hunk 6]
```diff
@@ -10,211 +10,157 @@
 module.exports = [
   // Separation Centre Referral - band 1,2 in custody
   {
-    uid: 'A4',
```
Comment (timharrison-moj) [inline server/config/controlsAndInterventions/controls/controls.js:13]: Suggest a new file - called disruptions.js  - and leave the original controls as they were.

File: server/config/controlsAndInterventions/controls/getControls.js (hunk 2 above)
Comment (timharrison-moj) [reply server/config/controlsAndInterventions/controls/getControls.js:11]: Perhaps  different file - getDisruptions.js, leaving original controls as is?

File: server/config/controlsAndInterventions/controls/getControls.test.js [hunk 7]
```diff
@@ -1,74 +1,74 @@
 const R = require('ramda')
 const controls = require('./getControls')
 
-test('it should provide a mandatory Separation Centre Referral control for a non-TACT band 1 offender in custody', () => {
```
Comment (timharrison-moj) [inline server/config/controlsAndInterventions/controls/getControls.test.js:4]: Suggest a new test file getDisruptions.test.js and leave original file as is

File: server/data/pathfinderClient.js [hunk 8]
```diff
@@ -719,9 +720,9 @@ const getCimControls = async (nominalId, riskBandId, needsId, type, client = db)
              override_by AS "overrideBy",
              TO_CHAR(override_time, 'DD/MM/YYYY HH24:MI:SS') AS "overrideTime"
            from cim_controls
-            where nominal_id = $1 and risk_band_id = $2 and needs_id = $3 and control_type = $4
```
Comment (timharrison-moj) [inline server/data/pathfinderClient.js:722]: This is ok as is.

File: server/routes/controlsAndInterventions.js [hunk 9]
```diff
@@ -46,17 +46,19 @@ module.exports = function ctrlAndIntRoutes({ offenderService, pathfinderService,
    * @param controlType
    * @returns {string}
    */
+  // eslint-disable-next-line no-unused-vars
```
Comment (timharrison-moj) [inline server/routes/controlsAndInterventions.js:49]: Suggest different file for disruptions route and leave this one as is.

File: server/routes/controlsAndInterventions.test.js [hunk 10]
```diff
@@ -166,7 +166,7 @@ describe('controlsAndInterventions', () => {
     }, 10000)
   })
 
-  describe('GET for controls with override reasons available as hidden modals through a view rationale link', () => {
```
Comment (timharrison-moj) [inline server/routes/controlsAndInterventions.test.js:169]: Different file for disruption route tests? Leave this one?

File: server/routes/nominal.js [hunk 11]
```diff
@@ -18,6 +18,10 @@ module.exports = ({ offenderService, pathfinderService, documentService, cimServ
   get('/:nominalId/cim/controls', ctrlAndIntRoutes.getControls())
   post('/:nominalId/cim/controls', ctrlAndIntRoutes.postControls())
 
+  // Disruptions
+  get('/:nominalId/disruptions/add', ctrlAndIntRoutes.getControls())
+  post('/:nominalId/disruptions/add', ctrlAndIntRoutes.postControls())
+
```
Comment (timharrison-moj) [inline server/routes/nominal.js:24]: OK as is.

File: server/routes/nominal.test.js [hunk 12]
```diff
@@ -516,7 +516,7 @@ describe('nominal routes', () => {
           expect(res.text).toContain('Date of Birth')
           expect(res.text).toContain('Status and reason for inclusion')
           expect(res.text).toContain('Meetings')
-          expect(res.text).toContain('Actions, controls and interventions')
+          expect(res.text).toContain('Disruptions')
```
Comment (timharrison-moj) [inline server/routes/nominal.test.js:519]: OK as is.

File: server/services/cimService.js [hunk 13]
```diff
@@ -26,12 +26,12 @@ module.exports = function createCimService({ pathfinderClient, notificationServi
     }
 
     // Get the controls that are already present in the table for this nominal, risk band and needs id
-    const allControls = await pathfinderClient.getCimControls(nominalId, riskBandId, needsId, 'control')
+    const allControls = await pathfinderClient.getCimControls(nominalId, null, null, 'control')
     const controlsUpdateMethod = isNilOrEmpty(allControls)
-      ? insertAllControlsAndInterventions(configItems, 'control', nominalId, riskBandId, needsId, createdBy)
-      : alignControlsAndInterventions(allControls, configItems, 'control', nominalId, riskBandId, needsId, createdBy)
+      ? insertAllControlsAndInterventions(configItems, 'control', nominalId, null, null, createdBy)
+      : alignControlsAndInterventions(allControls, configItems, 'control', nominalId, null, null, createdBy)
```
Comment (timharrison-moj) [inline server/services/cimService.js:32]: Bit confusing this one - maybe take same approach to copy things relating to controls to a disruptions equivalent, and leaving as much of the original controls stuff in place as we can?

File: server/services/cimService.test.js [hunk 14]
```diff
@@ -41,7 +41,7 @@ afterEach(() => {
 })
 
 describe('cimServiceTest', () => {
-  test('it should generate and save a list of controls when first requested for a nominal', async () => {
```
Comment (timharrison-moj) [inline server/services/cimService.test.js:44]: Probably have separate tests for disruptions and leave the originals relating to controls in place?

File: server/views/pages/cim/controls.njk [hunk 15]
```diff
@@ -31,25 +31,26 @@
 {% from "../../partials/breadCrumb.njk" import breadCrumb %}
 
 {% block beforeContent %}
```
Comment (timharrison-moj) [inline server/views/pages/cim/controls.njk:33]: Separate template for disruptions? Leaving controls.njk as is.

File: server/views/partials/createControlsTable.njk [hunk 16]
```diff
@@ -26,7 +26,7 @@ Data supplied:
     <tr class="govuk-table__row">
         <th scope="col" class="govuk-table__header">UID</th>
         <th scope="col" class="govuk-table__header">Name of {{ controlType }}</th>
-        <th scope="col" class="govuk-table__header">Banding outcome</th>
+        {# <th scope="col" class="govuk-table__header">Banding outcome</th> #}
         <th scope="col" class="govuk-table__header">To be implemented?</th>
```
Comment (timharrison-moj) [inline server/views/partials/createControlsTable.njk:30]: Probably a separate macro for disruptions, leaving the original controls one as is?

**Comment by connormaglynn** [comment]
Revised in another branch

**Pull request closed without merging**